velox/functions/lib/Re2Functions.h

```diff
@@ -30,33 +30,128 @@ namespace facebook::velox::functions {
-  // otherwise.
-  size_t length;
-  // Contains the unescaped fixed pattern in patterns of kind kFixed, kPrefix,
-  // kSuffix and kSubstring.
-  std::string fixedPattern = "";
+/// Kind of sub-pattern.
+enum class SubPatternKind {
+  /// e.g. '___'.
+  kSingleCharWildcard,
+  // e.g. '%%'.
+  kAnyCharsWildcard,
+  // e.g. 'abc'.
+  kLiteralString
 };
+
+class PatternMetadata {
+ public:
+  static PatternMetadata generic();
+
+  static PatternMetadata atLeastN(size_t length);
+
+  static PatternMetadata exactlyN(size_t length);
+
+  static PatternMetadata fixed(const std::string& fixedPattern);
+
+  static PatternMetadata relaxedFixed(
+      const std::string& fixedPattern,
+      const std::vector<SubPatternKind>& subPatternKinds,
+      const std::vector<std::pair<size_t, size_t>>& subPatternRanges);
+
+  static PatternMetadata prefix(const std::string& fixedPattern);
+
+  static PatternMetadata relaxedPrefix(
+      const std::string& fixedPattern,
+      const std::vector<SubPatternKind>& subPatternKinds,
+      const std::vector<std::pair<size_t, size_t>>& subPatternRanges);
+
+  static PatternMetadata suffix(const std::string& fixedPattern);
+
+  static PatternMetadata relaxedSuffix(
+      const std::string& fixedPattern,
+      const std::vector<SubPatternKind>& subPatternKinds,
+      const std::vector<std::pair<size_t, size_t>>& subPatternRanges);
+
+  static PatternMetadata substring(const std::string& fixedPattern);
+
+  const PatternKind patternKind() const {
+    return patternKind_;
+  }
+
+  const size_t length() const {
+    return length_;
+  }
+
+  const SubPatternKind subPatternKind(size_t index) const {
+    return subPatternKinds_[index];
+  }
+
+  const size_t numSubPatterns() const {
+    return subPatternRanges_.size();
+  }
+
+  const size_t subPatternStart(size_t index) const {
+    return subPatternRanges_[index].first;
+  }
+
+  const size_t subPatternLength(size_t index) const {
+    return subPatternRanges_[index].second;
+  }
+
+  const std::string& fixedPattern() const {
+    return fixedPattern_;
+  }
+
+ private:
+  PatternMetadata(
+      PatternKind patternKind,
+      size_t length,
+      std::string fixedPattern,
+      std::vector<SubPatternKind> subPatternKinds,
+      std::vector<std::pair<size_t, size_t>> subPatternRanges);
+
+  PatternKind patternKind_;
+
+  /// Contains the length of the unescaped fixed pattern for patterns of kind
+  /// k[Relaxed]Fixed, k[Relaxed]Prefix, k[Relaxed]Suffix and
+  /// k[Relaxed]Substring. Contains the count of wildcard character '_' for
+  /// patterns of kind kExactlyN and kAtLeastN. Contains 0/// otherwise.
+  size_t length_;
+
+  /// Contains the fixed pattern in patterns of kind k[Relaxed]Fixed,
+  /// k[Relaxed]Prefix, k[Relaxed]Suffix and k[Relaxed]Substring.
+  std::string fixedPattern_;
+
+  /// Contains the sub-pattern kinds for k[Relaxed]Xxx patterns.
+  std::vector<SubPatternKind> subPatternKinds_;
```

Comment:
Do we need these? We can only have single wilfcards and fixed strings. It seems sufficient to track ranges of fixed strings, no?

Reply:
We need to which range need to do exact match(e.g. 'abc'), which range only need to skip a certain length of string(e.g. '___', wildcards), only store range info is not eough?

Reply:
I'm thinking that there are only 2 possibilities: either skip or match. Hence, we can list only the ranges to match and assume the rest is to skip.

[[1, 3], [5, 4], [10, 7]] + pattern length 20 means

make sure there are at least 20 chars, then
skip to 1
match 3 chars
skip to 5
match 4 chars
skip to 10
match 7 chars

Would that work?